qiskit/transpiler/passes/optimization/optimize_annotated.py

```diff
@@ -148,17 +167,220 @@ def _canonicalize(self, dag) -> Tuple[DAGCircuit, bool]:
+            if len(preds) == 0:
+                for q in node.qargs:
+                    q_to_f[q] = node
+                    unchecked_qubits.add(q)
+            succs = dag.op_successors(node)
+            out_degree[node] = len(succs)
+            if len(succs) == 0:
+                for q in node.qargs:
+                    q_to_b[q] = node
+                    unchecked_qubits.add(q)
+
+        # iterate while there is a possibility to find more inverse pairs
+        while len(unchecked_qubits) > 0:
+            to_check = unchecked_qubits.copy()
+            unchecked_qubits.clear()
+
+            # For each qubit q, check whether the gate at the front of the circuit that involves q
+            # and the gate at the end of the circuit that involves q are inverse
+            for q in to_check:
+
+                if (fn := q_to_f.get(q, None)) is None:
+                    continue
+                if (bn := q_to_b.get(q, None)) is None:
+                    continue
+
+                # fn or bn could be already collected when considering other qubits
+                if fn in processed_nodes or bn in processed_nodes:
+                    continue
+
+                # it is possible that the same node is both at the front and at the back,
+                # it should not be collected
+                if fn == bn:
+                    continue
+
+                # have been checked before
+                if (fn, bn) in checked_node_pairs:
+                    continue
+
+                # fast check based on the arguments
+                if fn.qargs != bn.qargs or fn.cargs != bn.cargs:
+                    continue
+
+                # in the future we want to include a more precise check whether a pair
+                # of nodes are inverse
+                if fn.op == bn.op.inverse():
+                    # update q_to_f, q_to_b maps
+                    for q in fn.qargs:
+                        q_to_f[q] = None
+                    for q in bn.qargs:
+                        q_to_b[q] = None
+
+                    # see which other nodes become at the front and update information
+                    for node in dag.op_successors(fn):
+                        if node not in processed_nodes:
+                            in_degree[node] -= 1
+                            if in_degree[node] == 0:
+                                for q in node.qargs:
+                                    q_to_f[q] = node
+                                    unchecked_qubits.add(q)
+
+                    # see which other nodes become at the back and update information
+                    for node in dag.op_predecessors(bn):
+                        if node not in processed_nodes:
+                            out_degree[node] -= 1
+                            if out_degree[node] == 0:
+                                for q in node.qargs:
+                                    q_to_b[q] = node
+                                    unchecked_qubits.add(q)
```

Comment:
We don't discard the other `qargs` in an inverse pair out of `to_check`, which _might_ cause interesting edge-case interactions if we have a pair `A(0, 1) . B(0, 1) . <...> . B(0, 1) . A(0, 1)` or something - we might strip off the `A(0, 1)` pair while examining `0`, but then _also_ strip off the `B` pair while examining `1`. I'm not clear that that would actually cause any problems (what if A and B shared some but not all qargs?), but at a minimum it would be good to get some tests of that in.

Fwiw, if we wanted to sidestep it, then having the `for q in to_check` loop be something like:
```python
while to_check:
    qubit = to_check.pop()
    if fn == bn.inverse():
        for q in fn.qargs:
            to_check.discard(q)
```
could get around that.

Reply:
Sure, this can happen, but should be fine since the nodes get output in the correct order, so in your example we will add `A(0, 1)` first and `B(0, 1)` second. In fact, we probably even want this to happen to reduce the total number of main iterations (though probably it does not really matter). I will add more tests.